### ISSUE TITLE
Tinkering With a Golang Rewrite

### ISSUE THREAD

**Issue by Zibbp** [Zibbp/Ceres#10]
Hey all,

I'm tinkering around with a Go rewrite of Ceres. No promises if it will come out of the "tinker" stage. With that said do you have any features you think should be added if I plan to move forward?

A couple features I'm thinking of:
- UI overhaul
- Improved queue system with step restarts.
  - Better view to understand what step the archive process in on.
  - Chat render failed? Click restart and it will restart the chat render only.
- Quality selection when archiving a VOD.
- *Maybe in the future* YouTube VOD/Live support?

Let me know what you think.

**Comment by andymcblane** [comment]
Hey there! 

A few things come to mind after using it for a couple months (thank you for making it!): 

* Ability to customize the TwitchCLI commands 
* Potentially post-processing steps? Currently I am using the API and watching for new archives, and then post processing them with ffmpeg to reduce size & I also am stitching the video vod and chat render together
* A way to queue videos but only have them processed one (or two, three..) at a time (though live videos would still be downloaded). This would mean if I had 10 videos to archive, I could add them to the queue all at the same time. It might just be my setup though, I have about 150GB allocated to my Docker image and Ceres will crash if I add more than 2-3, ~6 hour VODs to archive

**Comment by Zibbp** [comment]
Hey,

Great suggestions. I like the queue one a lot. I've found that if I have more than 2 VODs in the queue that Twitch will at time rate limit me on either the video or chat download causing a lot of issues. I will make note if this one.

As for customizing the TwitchCLI commands. Did you have any in mind? Are you talking about the chat render options, which I know there are many of them.

Finally, about postprocessing. This would be difficult. With ffmpeg are you compressing the VOD quality so it's a lower file size? I know that these VODs can be quite large in size considering the average bitrate is between 6mb/s and 8mb/s. And by stitching the vod and chat render together, are you placing them side by side and rendering them? So the resolution is `VodWith+ChatWidth`?

I saw you were using the API. Any improvements you would like seen to the API. I was thinking about adding API keys that have a longer life duration. The issue with that is I don't plan on using any persistent sessions so you would not be able to revoke the API keys once generated until the time ran out.

**Comment by soleny894** [comment]
Hi,

Great news!

Here's my suggestions:
- Remake the admin interface for easy access: 
  - Merge "Manage Channel" and "Manage Live Channels" by adding a toggle for live archiving
- Being able to turn off user registration
- Multi VOD selection (ex: bulk deletion)
- Button or periodic check to update stream avatar (not sure if that's already the case)
- Check for stream live status (if container crash or restart, the channel is considered live indefinitely and don't download any new VOD automatically)

I'll edit this post if I find any other idea.

**Comment by Zibbp** [comment]
@soleny894 Thanks for the suggestions. I plan on creating a new user interface with many improvements, I'll be sure to include these suggestions.

> Button or periodic check to update stream avatar (not sure if that's already the case)

Are you talking about a button to update the VOD of the stream's thumbnail? I can create that as a button. I can also it as a function to run when the live stream archive finishes.

> Check for stream live status (if container crash or restart, the channel is considered live indefinitely and don't download any new VOD automatically)

When the X minute channel check is ran I can set any channels in the DB to not live is the Twitch API returns not live.

**Comment by Zibbp** [comment]
Question  - Currently in Ceres, the Twitch VOD/Stream ID is used as the primary key field meaning to query a VOD you would use the Twitch VOD Id `/api/v1/vods/1512811472`. I'm planning on moving the primary key field to a randomly generated ID meaning you wouldn't be able to access a VOD using the Twitch ID. It would look like so `/api/v1/vods/647ea39d-9519-4742-aa63-9516112e0f57`. I would make it so you could still filter in the admin panel by Twitch IDs but querying the VOD via the API or Frontend would require the random ID. 

TLDR: Does anyone use / care about the Twitch VOD/Stream ID other than starting an archive?

**Comment by soleny894** [comment]
> Are you talking about a button to update the VOD of the stream's thumbnail? I can create that as a button. I can also it as a function to run when the live stream archive finishes.

This can be a good thing to implement too. I was talking about the user avatar that show when you go in the "Channels" menu at the top. I'm not sure if it updates automatically already in this version.

> When the X minute channel check is ran I can set any channels in the DB to not live is the Twitch API returns not live.

Good thing. Sometime my containers are restarted and that keep the "live" status on certain channel, if I don't check by myself I can skip a few live stream archiving.

> Does anyone use / care about the Twitch VOD/Stream ID other than starting an archive?

No use for me personally.

**Comment by Zibbp** [comment]
> This can be a good thing to implement too. I was talking about the user avatar that show when you go in the "Channels" menu at the top. I'm not sure if it updates automatically already in this version.

It does not automatically update. If you would like to change it, simply replace the `channelName.jpg` where you VODs are located. `/mnt/vods/CHANNEL/CHANNEL.jpg` I could at a function that updates the channel avatar. It probably wouldn't be used often as I assume streamers don't update their channel avatar often.

**Comment by andymcblane** [comment]
> As for customizing the TwitchCLI commands. Did you have any in mind? Are you talking about the chat render options, which I know there are many of them.

Yep - I see it's currently rendering the chat out at 1440p and it's these types of things I'd like to adjust 

> Finally, about postprocessing. This would be difficult. With ffmpeg are you compressing the VOD quality so it's a lower file size? I know that these VODs can be quite large in size considering the average bitrate is between 6mb/s and 8mb/s. And by stitching the vod and chat render together, are you placing them side by side and rendering them? So the resolution is VodWith+ChatWidth?

Yeah I am reducing the quality - I likely will still use Ceres to get the best quality and then reduce size down based on type of content (e.g a speedrun would need less quality than a programming stream). 

Re the stitching, yeah the resulting width is `VodWidth + ChatWidth`. Unsure if other people would want this (especially as the Ceres UI displays side by side). 

No suggestions for the API - perfect already as I see it. I have no use for access tokens or anything like that. 

Few other things have come to mind:
* Resume video playback - e.g if I watch an hour of a vod, close the tab and then come back, it would be awesome if it prompted to resume play or start from beginning. 
* Potentially adding variable playback speed to the video player? 

> TLDR: Does anyone use / care about the Twitch VOD/Stream ID other than starting an archive?

Nope, though would be nice to store this information in the db / filename of the vod.

**Comment by Zibbp** [comment]
> Yep - I see it's currently rendering the chat out at 1440p and it's these types of things I'd like to adjust

That's easy enough to accomplish. I can make a page in the admin panel allowing lower level config changes such as that.

I'm not sure about the post-processing. That might be something you will have to implement yourself. Currently when a VOD is archived a webhook is sent out. You could utilize that webhook to detect when to run ffmpeg against the vod.

> - Resume video playback - e.g if I watch an hour of a vod, close the tab and then come back, it would be awesome if it prompted to resume play or start from beginning.
> - Potentially adding variable playback speed to the video player?

Good quality-of-life features that shouldn't be too bad to implement.

> Nope, though would be nice to store this information in the db / filename of the vod.

I plan on keeping the Twitch VOD ID in a `ext_id` column for display on the frontend. Working theory for filenames is `{UUID}-{twitchID}-{Video/Chat/Info.ext}`. This will allow you to easily check if a vod is archived via the Twitch ID or internal Ceres ID.

**Comment by Zibbp** [comment]
Hey,

If you would like to further suggest enhancements or would like to follow the rewrite please take a look at the below links.
https://github.com/Zibbp/ganymede/issues
https://github.com/users/Zibbp/projects/2/views/1

Feel free to continue suggesting enhancements or if I missed any, create a new issue in the `Ganymede` repository.

**Comment by Zibbp** [comment]
I did some thinking about the new queue system and how to implement the frontend UI. I will probably attempt to design a similar interface to Github Actions. 
![image](https://user-images.githubusercontent.com/21207065/176315643-06bdbcc8-c27e-47c5-8be3-c181b58182e8.png)
It will include different states for each step (success, processing, waiting, and error). Clicking on the video/chat download or chat render will take you to the logs. A restart button will be included next to each step allowing for a restart of any step.

Current steps:
**Vod**
create_folder
thumbnail_download
save_info

**Video**
video_download
video_move

**Chat**
chat_download
chat_render
chat_move

**Comment by Zibbp** [comment]
Hey all,

I've released a beta of the Ceres rewrite, Ganymede. Information can be found in this discussion https://github.com/Zibbp/ganymede/discussions/26 if you would like to test it out.

**Comment by Zibbp** [comment]
Ganymede release candidate v1 has been released. Visit the [Ganymede repository](https://github.com/Zibbp/ganymede) for further information.

Find any issues? Open an issue in the Ganymede repository.
